### PR TITLE
Add Support for deployment slots in deployment scripts

### LINKED ISSUE #428: Support for deployment slots in webapi and Preview Environments in webapp
Currently the azure deployment scripts do not support a deployment slot parameter. This feature will be beneficial for those who wish to deploy multiple versions of chat-copilot without standing up an entire new set of resources.

Both commands 'az webapp deploy' and 'az webapp config' used by deploy-webapi support the -slot parameter which defaults to 'Production' if not specified. An optional parameter will be made with a default value of Production. Following similar behavior in the deployment scripts the deployment slot would be created if it does not exist by using `az webapp deployment slot list` and `az webapp deployment slot create`

The deploy-webapp script pulls the webapi URL from the resource group deployment at `$deployment.properties.outputs.webapiUrl.value` this value may not be correct if the intent is to use additional deployment slots. To support the change in webapi URL created by the deployment slot an optional parameter in the deploy-webapp script will be created allowing the user to specifiy the url of the webapi. However if a redeployment of the webapp is required in order to access the webapi at a new URL it would be simpler to perform a slot swap on the webapi web app instead. This leads into the requirement for support of swa preview environments.

Static Web Apps do not support the use of deployment slots and instead use Preview Environments. `swa deploy` supports the use of `--env [environment]`. My understanding is that this will create a new preview environment with a URL in the form of `<DEFAULT_HOST_NAME>-<BRANCH_OR_ENVIRONMENT_NAME>.<LOCATION>.azurestaticapps.net`. An optional parameter in deploy-webapp will be created to support the use of the Preview Environments.

### PR DESCRIPTION
**Pull request by arnoldblake**
### Motivation and Context

This pull request introduces an enhancement to the deploy-webapi and deploy-webapp scripts by enabling the optional use of Azure App Service Deployment Slots and Static Web App Preview Environments respectively. This feature addresses the challenge of achieving zero-downtime deployments, ensuring that our application remains accessible and functional during updates. Deployment slots allow us to validate application changes in a staging environment that mirrors production, facilitating extensive testing and verification before the swap. This guarantees that only thoroughly vetted and approved changes reach the live environment, enhancing our deployment confidence and application reliability. Moreover, in the rare case of an issue arising post-deployment, the slots provide an instant rollback mechanism, minimizing disruption and ensuring continuity of service for our users.

### Description

resolves #428 

Both scripts were modified to allow an optional parameter for either a deployment slot or preview environment. The scripts behavior is unchanged if these parameters are not present. 

deploy-webapi
Code was introduced which checks for the existence of a deployment slot. Following other script behavior if the deployment slot does not exist it is created. 

deploy-webapp
Modified the command swa deploy to support the environment parameter which defaults to production if not present. Added a loop to check each preview environment URL is present in the CORS and AAD app registration.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [Contribution Guidelines](https://github.com/microsoft/chat-copilot/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/chat-copilot/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: scripts/deploy/deploy-webapi.ps1 [hunk 1]
```diff
@@ -56,8 +60,39 @@ if ($LASTEXITCODE -ne 0) {
     exit $LASTEXITCODE
 }
 
+# Check if DeploymentSlot parameter was passed
+if ($DeploymentSlot) {
+
+    Write-Host "Checking if slot $DeploymentSlot exists for '$webappName'..."
+    $availableSlots = az webapp deployment slot list --resource-group utverse-devl --name app-utverse-devl-webapi | ConvertFrom-JSON | Select-Object -Property Name
```
Comment (glahaye) [inline scripts/deploy/deploy-webapi.ps1:67]: utverse-devl???  That should be replaced by the param to this script! Same at line 80.

File: scripts/deploy/deploy-webapi.sh [hunk 2]
```diff
@@ -12,6 +12,7 @@ usage() {
     echo "  -s, --subscription SUBSCRIPTION         Subscription to which to make the deployment (mandatory)"
     echo "  -rg, --resource-group RESOURCE_GROUP    Resource group name from a 'deploy-azure.sh' deployment (mandatory)"
     echo "  -p, --package PACKAGE_FILE_PATH         Path to the WebAPI package file from a 'package-webapi.sh' run (default: \"./out/webapi.zip\")"
+    echo "  -o, --slot DEPLOYMENT_SLOT                      Name of the target web app deployment slot"
```
Comment (glahaye) [inline scripts/deploy/deploy-webapi.sh:15]: Please line this up with the other descriptions above

File: scripts/deploy/deploy-webapi.sh [hunk 3]
```diff
@@ -80,14 +86,48 @@ fi
...
+        if [ "$SLOT" == "$DEPLOYMENT_SLOT" ]; then
+            SLOT_EXISTS=true
+            break
+        fi
+    done
+
+    # If slot does not exist, create it
+    if [ "$SLOT_EXISTS" == "false" ]; then
+        echo "Slot $DEPLOYMENT_SLOT does not exist, creating..."
+        az webapp deployment slot create --slot $DEPLOYMENT_SLOT --resource-group $RESOURCE_GROUP --name $WEB_APP_NAME > /dev/null 2>&1
```
Comment (glahaye) [inline scripts/deploy/deploy-webapi.sh:115]: Are you sure you want to redirect BOTH stdout and stderr to null?

File: scripts/deploy/deploy-webapp.sh [hunk 4]
```diff
@@ -136,42 +145,50 @@ if [ $? -ne 0 ]; then
...
 popd
 
-ORIGIN="https://$WEB_APP_URL"
-echo "Ensuring origin '$ORIGIN' is included in CORS origins for webapi '$WEB_API_NAME'..."
-CORS_RESULT=$(az webapp cors show --name $WEB_API_NAME --resource-group $RESOURCE_GROUP --subscription $SUBSCRIPTION | jq '.allowedOrigins | index("$ORIGIN")')
-if [[ "$CORS_RESULT" == "null" ]]; then
-    echo "Adding CORS origin '$ORIGIN' to webapi '$WEB_API_NAME'..."
-    az webapp cors add --name $WEB_API_NAME --resource-group $RESOURCE_GROUP --subscription $SUBSCRIPTION --allowed-origins $ORIGIN
-fi
+APPNAME="swa-utverse-devl"
```
Comment (glahaye) [inline scripts/deploy/deploy-webapp.sh:156]: This hardcoded value should not be here...

**Review by glahaye — CHANGES_REQUESTED** [review]
Some hardcoded values were left in the scripts - They should be replaced with the appropriate parameter value

**Review by glahaye — DISMISSED** [review]

**Comment by glahaye** [comment]
Just sync with latest and resolve the conflicts and this PR is good to go

**Comment by arnoldblake** [comment]
@microsoft-github-policy-service agree

**Comment by arnoldblake** [comment]
Will the commits be squashed when they are merged?

**Comment by glahaye** [comment]
> Will the commits be squashed when they are merged? Should I squash them myself? How would that impact the pull request?

They will be squashed. Please don't waste time doing that yourself! :)

**Review by glahaye — APPROVED** [review]

**Pull request merged**